src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,361 @@
+    fun_transform(fun, x::T) = fun(x)
+    map(i -> fun_transform(fun, T(ntuple(j -> cols[j][i], length(cols)))), 1:n)
+end
+
+function select_transform!(nc::Union{Pair{Int, Pair{ColRename, Symbol}},
+                                     Pair{<:Union{Int, AbstractVector{Int}},
+                                          <:Pair{<:Base.Callable, Symbol}}},
+                           df::DataFrame, newdf::DataFrame,
+                           transformed_cols::Dict{Any, Any}, copycols::Bool)
+    newname = last(last(nc))
+    if !isnothing(transformed_cols[newname])
+        @assert !hasproperty(newdf, newname)
+    end
+    col_idx = first(nc)
+    if nc isa Pair{Int, Pair{ColRename, Symbol}}
+        newdf[!, newname] = copycols ? df[:, col_idx] : df[!, col_idx]
+    elseif nc isa Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+        newdf[!, newname] = first(last(nc)).(df[!, col_idx])
+    elseif nc isa Pair{<:AbstractVector{Int}, <:Pair{<:Base.Callable, Symbol}}
+        if length(col_idx) == 0
+            newdf[!, newname] = [first(last(nc))() for _ in axes(df, 1)]
+        else
+            cols = ntuple(i -> _columns(df)[col_idx[i]], length(col_idx))
+            colnames = ntuple(i -> _names(df)[col_idx[i]], length(col_idx))
+            newdf[!, newname] = select_transform_helper(TypeHolder{NamedTuple{colnames,
+                                                                              Tuple{eltype.(cols)...}}}(),
+                                                        cols, first(last(nc)), nrow(df))
+        end
+    else
+        throw(ErrorException("code should never reach this branch"))
+    end
+    transformed_cols[newname] = nothing
+end
+
+"""
+    select!(df::DataFrame, inds...)
+
+Mutate `df` in place to retain only columns specified by `inds...` and return it.
+
+Arguments passed as `inds...` can be any index that is allowed for column indexing
+provided that the columns requested in each of them are unique and present in `df`.
+In particular, regular expressions, `All`, `Between`, and `Not` selectors are supported.
+
+Column renaming and transformations are supported.
+The syntax for column renaming is `old_column=>new_column_name`.
+The syntax for column transformations is `old_column=>fun=>new_column_name`.
+`new_column_name` must be a `Symbol`.
+If `old_column` is a `Symbol` or an integer then and `fun` must be callable
+that is applied row by row to the values of `old_column`.
+Otherwise `old_column` can be any column indexing syntax, but in this case `fun`
+will be passed `DataFrameRows` selected by `old_column`.
+
+Column transfomation syntax also supports a short `old_column=>fun` form, in which
+case `new_column_name` is automatically generated by joining `olcd_column` with `fun` name
+with `_`.
+
+If more than one argument is passed then duplicates are accepted except for
+column renaming and transformation operations, where it is not alloweded to rename/transform
```

Comment:
The part after "where" isn't clear: if that's just a repetition of the first part, better remove it. If it means that you cannot do `:x => :x`, then that's a different issue (is that the case?).